Step1

hjlee83 opened the pull request:
잘 부탁드립니다.

hojinDev committed:
feat(dsl): add custom DSL for building person objects

Person DSL supports dynamic definition of name, company, skills, and languages.
Introduced flexible builders to simplify creation and testing of `Person` objects.

Added comprehensive tests to validate the DSL functionality and ensure correctness.

hyogeonkim committed:
feat(dsl): add language block

hyogeonkim committed:
doc: README.md

hyogeonkim committed:
feat: input playerNames

hyogeonkim committed:
feat: Deck, Card

hyogeonkim committed:
etc: formatting

hojinDev committed:
feat(readme): update checklist for blackjack requirements

aligned checklist with current implementation status by marking completed items
such as card shuffle and initial distribution. clarified pending tasks for better planning
and progress tracking.

hojinDev committed:
feat(readme): update checklist for blackjack requirements

aligned checklist with current implementation status by marking completed items
such as card shuffle and initial distribution. clarified pending tasks for better planning
and progress tracking.

hojinDev committed:
feat(blackjack): implement dealer and player mechanics with tests

added `Dealer` and `Player` classes inheriting from `Participant`, encapsulating
core blackjack logic. implemented dealing decisions for dealer based on
card value conditions. updated `Rank` enum to expose `score` for calculations.

introduced `OutView` to display game states such as initial distribution
and player card lists. added tests for `Dealer` and `Player` to validate
functionality.

updated README checklist to reflect progress on completed tasks.

hojinDev committed:
feat(blackjack): implement dealer and player mechanics with tests

added `Dealer` and `Player` classes inheriting from `Participant`, encapsulating
core blackjack logic. implemented dealing decisions for dealer based on
card value conditions. updated `Rank` enum to expose `score` for calculations.

introduced `OutView` to display game states such as initial distribution
and player card lists. added tests for `Dealer` and `Player` to validate
functionality.

updated README checklist to reflect progress on completed tasks.

hojinDev committed:
feat(blackjack): calculate ace as 1 or 11 and refactor score logic

added `totalScore` method in `Participant` to handle flexible scoring for Ace as 1 or 11.
updated `Dealer` to utilize `totalScore` for cleaner logic and improved readability.

adjusted README checklist to reflect progress on Ace scoring and dealer card addition tasks.

hojinDev committed:
feat(blackjack): calculate ace as 1 or 11 and refactor score logic

added `totalScore` method in `Participant` to handle flexible scoring for Ace as 1 or 11.
updated `Dealer` to utilize `totalScore` for cleaner logic and improved readability.

adjusted README checklist to reflect progress on Ace scoring and dealer card addition tasks.

hojinDev committed:
feat(blackjack): enhance rank display and input validation

updated `Rank` enum to include display name for improved user readability.
enhanced `InputView` to ensure all player names are non-blank during input validation.

adjusted README to refine checklist formatting and align with recent updates.

hojinDev committed:
feat(blackjack): enhance rank display and input validation

updated `Rank` enum to include display name for improved user readability.
enhanced `InputView` to ensure all player names are non-blank during input validation.

adjusted README to refine checklist formatting and align with recent updates.

hojinDev committed:
feat(blackjack): improve game flow and output readability

renamed `OutView` to `OutputView` for consistent naming convention. added `toString` method in `Card` for better card display. introduced dealer's initial card visibility and player's decision functionality for additional cards.

updated `InputView` for flexible player input handling and validation. refined exception handling by displaying error messages for invalid inputs.

hojinDev committed:
feat(blackjack): improve game flow and output readability

renamed `OutView` to `OutputView` for consistent naming convention. added `toString` method in `Card` for better card display. introduced dealer's initial card visibility and player's decision functionality for additional cards.

updated `InputView` for flexible player input handling and validation. refined exception handling by displaying error messages for invalid inputs.

hyogeonkim committed:
feat: result score, record

hojinDev committed:
Merge remote-tracking branch 'men7627/step1' into step1

# Conflicts:
#	README.md
#	src/main/kotlin/blackjack/BlackjackApplication.kt
#	src/main/kotlin/blackjack/domain/Dealer.kt
#	src/main/kotlin/blackjack/domain/Participant.kt
#	src/main/kotlin/blackjack/domain/Player.kt
#	src/main/kotlin/blackjack/model/Card.kt
#	src/main/kotlin/blackjack/model/CardEnums.kt
#	src/main/kotlin/blackjack/view/InputView.kt
#	src/main/kotlin/blackjack/view/OutputView.kt
#	src/test/kotlin/blackjack/domain/DealerTest.kt
#	src/test/kotlin/blackjack/domain/PlayerTest.kt

hojinDev committed:
feat(blackjack): enhance deck functionality and improve input handling

updated `Deck` to use a mutable list for cards, enabling proper card removal and shuffle.
added exception handling when attempting to pop cards from an empty deck. improved unit tests
to validate deck behavior, ensuring uniqueness and proper handling of an empty deck.

refined `InputView` to utilize `retryInput` for better user input validation and error handling.

hojinDev committed:
refactor(blackjack): extract reusable methods and simplify logic

replace inline score calculation and ACE check with `score()` and `isAce()` methods in `Card`.
refactor `Participant` to utilize these methods for cleaner and more readable scoring logic.

simplify `players()` function in `BlackjackApplication` by removing redundant intermediate variable.
update tests to use `toString()` for unique identification of cards, ensuring consistent behavior.

hojinDev committed:
refactor(blackjack): simplify `Dealer` additional deal condition logic

replaced block body syntax in `isAdditionalDealCondition` with expression body for brevity.
removed redundant parentheses in the `Dealer` class constructor for cleaner code.

hojinDev committed:
refactor(blackjack): move constants to companion objects for better encapsulation

relocated `DEALER_NAME`, `DEALER_ADD_CARD_BASE_SCORE`, and `ACE_ADDITIONAL_SCORE` constants from `Constants.kt` to respective companion objects in `Dealer` and `Participant`.

improves encapsulation and makes the constants more contextually relevant to their usages.

hojinDev committed:
refactor(blackjack): simplify enums and scoring logic in `Player`

reduce boilerplate in `Suit`, `Rank`, and `RecordType` enums by converting to expression bodies. update `Player` scoring method to streamline logic and enhance readability.

improves code clarity, reduces redundancy, and aligns with previous refactoring patterns.

hojinDev committed:
refactor(blackjack): enforce dealer card limit and refactor card count logic

add card limit enforcement in `Dealer`, restricting cards to a maximum of 3. introduce `cardCount()` method in `Participant` to simplify card counting logic. update relevant test cases to validate dealer's card limit and improve test descriptions for clarity.

ensures adherence to game rules by preventing dealers from exceeding allowed card limits and streamlines card-related operations across participants.

hojinDev committed:
feat(blackjack): complete game result and player victory logic

README 업데이트로 참가자 결과 출력 및 승패 출력 작업 완료 상태 반영.
`InputView`에서 `retryInput` 사용을 통해 입력 검증 및 올바른

hojinDev committed:
feat(blackjack): implement game result and improve output readability

refactor `OutputView` to streamline message generation for game results and add new methods for final output, improving readability. introduce `Result`, `PlayerRecord`, and `DealerWinDrawLose` for structured handling of match outcomes.

enhances clarity and modularity for result calculation and final display by abstracting repeated logic and providing detailed information on dealer and players.

HyunGyu-Lee commented on src/main/kotlin/blackjack/BlackjackApplication.kt:
```diff
@@ -0,0 +1,78 @@
+    OutputView.initialDeal(players.map { it.name })
+
+    val participants = players + dealer
+    participants.forEach { participant ->
```
요건 취향/규칙의 영역인 것 같은데 저는 이 부분을 딜러/플레이어에서 분배하는 부분도 하나의 비즈니스 로직이라고 생각하고 별도 서비스로 분리를 해 볼 수 도 있을 것 같고 딜러 / 플레이어 분배 규칙이 바뀔수도 있으니까 이를 분리하는 것도 선택해볼수있는 옵션 같습니다! 
근데 지금 형태도 충분히 괜찮은것 같아요!

hjlee83 replied:
현규님 요고는 일단 이렇게 진행하겠습니다. 후후후 다음 스텝을 빨리 하고 파요.

HyunGyu-Lee commented on src/main/kotlin/blackjack/domain/Participant.kt:
```diff
@@ -0,0 +1,31 @@
+    private val _cards = mutableSetOf<Card>()
+
+    val cards: Set<Card>
+        get() = _cards
```
제가 확실하진 않지만 요렇게 cards 를 외부에서 획득하고 MutableSet 으로 캐스팅을 하면 수정이 가능할 수도 있을 것 같은데 
`_cards.toSet()` 과 같이 복사본을 제공하면 아예 수정 불가한 + 복사본이 제공되니 원본에 영향을 주지 않을 것 같습니다!

HyunGyu-Lee commented on src/main/kotlin/blackjack/domain/Enums.kt:
```diff
@@ -0,0 +1,38 @@
+
+enum class RecordType(val displayName: String) {
+    WIN("승"),
+    DRAW("무"),
```
오 무승부도 구현하셨군요..

HyunGyu-Lee commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -0,0 +1,33 @@
+    fun playerNames(): List<String> {
+        println("게임에 참여할 사람의 이름을 입력하세요.(쉼표 기준으로 분리)")
+
+        return InputUtils.retryInput {
```
요렇게 입력부 한번 감싸주신것도 너무 멋지네요.. 블랙잭 다음 단계 하게 되면 차용해보겠습니다

HyunGyu-Lee reviewed (changes requested):
고생많으셨습니다. 리뷰라기보단 제 감상평 + 소소한 의견 드린 것 같은데 가볍게 살펴봐주세요~!

hojinDev committed:
refactor(blackjack): ensure immutability in `cards` property of `Participant`

replace direct `_cards` reference with `.toSet()` for getter logic in `cards` property, enforcing immutability.

aligns with best practices by preventing external modifications to internal card state.

HyunGyu-Lee reviewed (approved):
고생 많으셨습니다~!
